test/attn_torch_function.py

```diff
@@ -58,8 +75,73 @@ def forward(ctx, q, k, v, b, causal, sm_scale, dropout_p, return_encoded_softmax
+                        dropout_p, philox_seed, philox_offset, encoded_softmax, causal,
+                        extargs)
+                try:
+                    ret = attn_fwd(*args)
+                except Exception as e:
+                    print(e)
+                    return hipError_t.hipErrorLaunchFailure, None
+                return ret, (o,)
+            def ipc_func(force_kernel_index):
+                shard = attn_extra_args.gpu_device
+                tune_worker = attn_extra_args.tune_worker
+                def factory():
+                    ipc_to_worker = torch.multiprocessing.Queue()
+                    ipc_from_worker = torch.multiprocessing.Queue()
+                    ipc_to_worker.cancel_join_thread()
+                    ipc_from_worker.cancel_join_thread()
+                    p = Process(target=ipc_attn_fwd,
+                                args=(ipc_to_worker, ipc_from_worker))
+                    p.start()
+                    return (ipc_to_worker, ipc_from_worker, p)
+                ipc_to_worker, ipc_from_worker, p = tune_worker.request_cached_gpukernel_process(ipc_attn_fwd, factory)
+                # print(f'{q.data_ptr()=:x}')
+                # print(f'{k.data_ptr()=:x}')
+                # print(f'{v.data_ptr()=:x}')
+                # print(f'{b.data_ptr()=:x}')
+                # print(f'{M.data_ptr()=:x}')
+                # print(f'{o.data_ptr()=:x}')
+                ipc_to_worker.put((q, k, v, b, sm_scale, M, o,
+                                   dropout_p, philox_seed, philox_offset, encoded_softmax, causal,
+                                   force_kernel_index, shard))
+                while p.is_alive():
+                    try:
+                        iret = ipc_from_worker.get(timeout=1)
+                        break
+                    except queue.Empty:
+                        # print(f'Process timeout {p.is_alive()=}')
+                        pass
+                # print(f'Process attn_fwd starting')
+                if not p.is_alive():
+                    # print(f'Process exitcode {p.exitcode}')
+                    tune_worker.invalid_gpukernel_process_cache(ipc_attn_fwd)
+                    p.join()
+                    ret = hipError_t.hipErrorLaunchFailure
+                else:
+                    ret = hipError_t.hipSuccess if iret == 0 else hipError_t.hipErrorLaunchFailure
+                # print(f'Process attn_fwd joined')
+                # print(f'Process exitcode {p.exitcode}')
+                return ret, (o,)
+            def func(extargs, is_testing):
+                # print(f'{is_testing=}')
+                if not is_testing:
+                    return sameprocess_func(extargs)
+                o.fill_(float('nan'))
+                return ipc_func(extargs.force_kernel_index)
+                # print(f'running attn_fwd with {extargs.force_kernel_index=}')
+            tuning_result = cpp_autotune(ExtraArguments, func,
```

Comment:
Any plan to expose this function (ExtraArguments) or is it only for tuning?

Reply:
This will be the part of public AOTriton API for possible further extensions.
For example, we may want to add a second hipStream in the ExtraArguments for bwd kernel, to run the dkdv and dq concurrently.